app/upgrades/signetlaunch/upgrades_test.go

```diff
@@ -0,0 +1,125 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	DummyUpgradeHeight = 5
+)
+
+type UpgradeTestSuite struct {
+	suite.Suite
+
+	ctx       sdk.Context
+	app       *app.BabylonApp
+	preModule appmodule.HasPreBlocker
+}
+
+func (s *UpgradeTestSuite) SetupTest() {
+	// add the upgrade plan
+	app.Upgrades = append(app.Upgrades, v1.Upgrade)
+
+	// set up app
+	s.app = app.Setup(s.T(), false)
+	s.ctx = s.app.BaseApp.NewContextLegacy(false, tmproto.Header{Height: 1, ChainID: "babylon-1", Time: time.Now().UTC()})
+	s.preModule = upgrade.NewAppModule(s.app.UpgradeKeeper, s.app.AccountKeeper.AddressCodec())
+
+	btcHeaderZero, err := app.SignetBtcHeaderZero(s.app.EncodingConfig().Codec)
+	s.NoError(err)
+
+	k := s.app.BTCLightClientKeeper
+	btclightclient.InitGenesis(s.ctx, s.app.BTCLightClientKeeper, btclighttypes.GenesisState{
+		Params:     k.GetParams(s.ctx),
+		BtcHeaders: []*btclighttypes.BTCHeaderInfo{btcHeaderZero},
+	})
+}
+
+func TestKeeperTestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+func (s *UpgradeTestSuite) TestUpgrade() {
+	oldHeadersLen := 0
+
+	testCases := []struct {
+		msg         string
+		pre_update  func()
+		update      func()
+		post_update func()
+	}{
+		{
+			"Test launch software upgrade gov prop",
+			func() {
+				allBtcHeaders := s.app.BTCLightClientKeeper.GetMainChainFrom(s.ctx, 0)
+				oldHeadersLen = len(allBtcHeaders)
+			},
+			func() {
+				// inject upgrade plan
+				s.ctx = s.ctx.WithBlockHeight(DummyUpgradeHeight - 1)
+				plan := upgradetypes.Plan{Name: v1.Upgrade.UpgradeName, Height: DummyUpgradeHeight}
+				err := s.app.UpgradeKeeper.ScheduleUpgrade(s.ctx, plan)
+				s.NoError(err)
+
+				// ensure upgrade plan exists
+				actualPlan, err := s.app.UpgradeKeeper.GetUpgradePlan(s.ctx)
+				s.NoError(err)
+				s.Equal(plan, actualPlan)
+
+				// execute upgrade
+				s.ctx = s.ctx.WithHeaderInfo(header.Info{Height: DummyUpgradeHeight, Time: s.ctx.BlockTime().Add(time.Second)}).WithBlockHeight(DummyUpgradeHeight)
+				s.NotPanics(func() {
+					_, err := s.preModule.PreBlock(s.ctx)
+					s.NoError(err)
+				})
+			},
+			func() {
+				// ensure the btc headers were added
+				allBtcHeaders := s.app.BTCLightClientKeeper.GetMainChainFrom(s.ctx, 0)
+
+				btcHeadersInserted, err := v1.LoadBTCHeadersFromData()
+				s.NoError(err)
+				lenHeadersInserted := len(btcHeadersInserted)
+
+				newHeadersLen := len(allBtcHeaders)
+				s.Equal(newHeadersLen, oldHeadersLen+lenHeadersInserted)
+
+				// ensure the headers were inserted as expected
+				for i, btcHeaderInserted := range btcHeadersInserted {
+					btcHeaderInState := allBtcHeaders[oldHeadersLen+i]
+
+					s.EqualValues(btcHeaderInserted.Hash.MarshalHex(), btcHeaderInState.Hash.MarshalHex())
+					s.EqualValues(btcHeaderInserted.Header.MarshalHex(), btcHeaderInState.Header.MarshalHex())
+					s.EqualValues(btcHeaderInserted.Height, btcHeaderInState.Height)
+
+					// TODO: check why work does not match
```

Comment:
@KonradStaniec by using [`k.InsertHeaders`](https://github.com/babylonlabs-io/babylon/blob/3c9ae6132c910d3ffed4c9d732bbcb68d0958ea3/app/upgrades/signetlaunch/upgrades.go#L107) to insert the BTC Headers, the work value does not match with the work retrieved from BTC https://github.com/babylonlabs-io/babylon/blob/3c9ae6132c910d3ffed4c9d732bbcb68d0958ea3/app/upgrades/signetlaunch/data_btc_headers.go#L3

Is that expected?

Reply:
uh good we cought it no.

So in structure:
```
    {
      "header": "00000020533b53ded9bff4adc94101d32400a144c54edc5ed492a3b26c63b2d686000000b38fef50592017cfafbcab88eb3d9cf50b2c801711cad8299495d26df5e54812e7914d5fae77031ecfdd0b00",
      "hash": "00000032bb881de703dcc968e8258080c7ed4a2933e3a35888fa0b2f75f36029",
      "height": 2,
      "work": "77414720"
    },
    {
      "header": "000000202960f3752f0bfa8858a3e333294aedc7808025e868c9dc03e71d88bb320000007765fcd3d5b4966beb338bba2675dc2cf2ad28d4ad1d83bdb6f286e7e27ac1f807924d5fae77031e81d60b00",
      "hash": "000000e8daac2a2e973ecaab46dc948181c638adecf5ae0fd5d3e13aa14364b0",
      "height": 3,
      "work": "77414720"
    },
```

work should be total work of BTC chain at each given height, not just work of given header.  So if we have two headers from this example the work of the second one should be `2 *  77414720`

Given that, maybe it is better if data to upgrade will contain raw headers ie.
```
   "header": "000000202960f3752f0bfa8858a3e333294aedc7808025e868c9dc03e71d88bb320000007765fcd3d5b4966beb338bba2675dc2cf2ad28d4ad1d83bdb6f286e7e27ac1f807924d5fae77031e81d60b00"
```

Then btc light client module will automatically calculates all the work for all the height and make sure all validations are passed as expected ?

Reply:
got it, makes sense 
I will update, that also requires modification at the generation of this headers at the `sid` btc indexer

Reply:
done at https://github.com/babylonlabs-io/babylon/pull/4/commits/e10c0324cae4bd87f0c6ec80fe044c39f5a89e5f